proposal/Naetw/README.md

```diff
@@ -0,0 +1,116 @@
+
+### The `Color` class
+
+This class stores colors (RGB) and provides a operation for writing a single pixel's color to the standard output stream in the PPM image format.
+
+### The `Ray` class
+
+This class describes a ray which can be thought of a formula $P(t) = A + tb$, where $A$ is the origin and `b` is the ray direction. Different $t$ moves the point along the ray to different positions.
+
+It also provides a method for generating the color seen in the direction of a ray.
+
+### The `HittableObject` class
+
+This is an abstract class used to represent any kind of object that can be hit by a ray. This should provide a hit (virtual?) function that tell the result of whether the given ray hits it.
+
+### The `Sphere` class
+
+> Inherited from class `HittableObject`
+
+This class describes a sphere which is often used in ray tracers since it's quite straightforward to know whether a ray hits it and it's useful for the start.
+
+It also provides a method that calculate the surface normal for the given hit point. This would be useful when we need to shade.
+
+### The `Camera` class
+
+I haven't had a thorough understanding of this component in ray tracing. It seems like a pinhole (observation point) which collects the rays we need and the ray tracer will use that to render the imagery.
+
+### The `Material` class
+
+This class describes the material which is going to be used as an attribute of an object and the key to generate different kinds of scattered ray.
+
+### Misc utility
+
+Ray tracer needs some math constants such as pi, infinity and some utility functions such as translation between degree and radian. This may forms a `Math` class or just a module that places all these hard to be classified stuffs.
+
+### Workflow
+
+1. Have a scene that the imagery would be generated at
+2. Have a "world" which describes objects that exist in the scene
+3. Use a camera to get the ray that we'd like to observe
+4. Use the ray got from step 3 to render the result
+
+## API description
+
+I may design a format for the user to describe the "world" which should be passed to the API (maybe called `generate_image` or something).
+
+I don't have too much time for this recently. I'll leave it until I clarify how much I'm going to extend the ray tracer under the constraint of available time in this semester.
+
+## Engineering infrastructure
+
+For different APIs, I plan to use different testing frameworks.
+
+- Public APIs (Python + C++)
+    - pytest
+- Private APIs (C++)
+    - SkyPat (used before) or Catch2
+
+As for documentation, I'm used to write the documentation first before actually writing some code if I have enough time. So, ideally, I would have a documentation which describes all the abstract information about my system. Besides, since this system is going to have a lot of formulas which is hard to be clearly represented by the code, I intend to write comments depict the formula in a more human-friendly manner for them.
```

Comment:
It's good to have documentation as you code.  It could be hard to write document before you have code, but it's good to try.  If you find you are spending too much time writing document, please switch to code as soon as possible.

Reply:
Sure! I'm not going to write a thorough document before I code. It's like an abstract description of code behavior (a simple design document). It helps me a lot for having some rough image of the system but it indeed takes considerable time. I'm still practicing to strike a balance on how much time should I spend for it. Thanks for your reminder!